src/autodiff.jl

```diff
@@ -382,6 +382,136 @@ function residual_jacobian!(arrays::Jacobian,
+        global dJ = zeros(Float64, length(F), length(arrays.map))
+    elseif isa(type, ControlJacobian)
+        arrays.x[1:nvbus] .= v_m
+        arrays.x[nvbus+1:nvbus+ninj] .= pinj
+        global dJ = zeros(Float64, length(F), length(arrays.map))
+    else
+        error("Unsupported Jacobian structure")
+    end
+    x = arrays.x
+    adj_x = zeros(Float64, length(x))
+    for i in 1:length(F)
+        F .= 0
+        adj_F .= 0.0
+        adj_F[i] = 1.0
+        adj_x .= 0.0
+        if isa(type, StateJacobian)
+            residual_adj_polar!(
+                F, adj_F,
+                view(x, 1:nvbus),
+                view(adj_x, 1:nvbus),
+                view(x, nvbus+1:2*nvbus),
+                view(adj_x, nvbus+1:2*nvbus),
+                ybus_re, ybus_im,
+                pinj, zeros(eltype(pinj), length(pinj)),
+                qinj, zeros(eltype(qinj), length(qinj)),
+                pv, pq, nbus
+            )
+            dJ[i,:] .= adj_x[arrays.map]
+        elseif isa(type, ControlJacobian)
+            residual_adj_polar!(
+                F, adj_F,
+                view(x, 1:nvbus),
+                view(adj_x, 1:nvbus),
+                v_a, zeros(eltype(v_a), length(v_a)),
+                ybus_re, ybus_im,
+                view(x, nvbus+1:nvbus+ninj),
+                view(adj_x, nvbus+1:nvbus+ninj),
+                qinj, zeros(eltype(qinj), length(qinj)),
+                pv, pq, nbus
+            )
+            dJ[i,:] .= adj_x[arrays.map]
+        else
+            error("Unsupported Jacobian structure")
+        end
+    end
+    return sparse(dJ)
+end
+
+function residual_hessian!(arrays::Jacobian,
+                             residual_adj_polar!,
+                             lambda,
+                             v_m, v_a, ybus_re, ybus_im, pinj, qinj, pv, pq, ref, nbus,
+                             type::AbstractJacobian)
+    x = arrays.x
+    nvbus = length(v_m)
+    ninj = length(pinj)
+    t1sx = arrays.t1sx
+    adj_t1sx = similar(t1sx)
+    t1sF = arrays.t1sF
+    adj_t1sF = similar(t1sF)
+    if isa(type, StateJacobian)
+        x[1:nvbus] .= v_m
+        x[nvbus+1:2*nvbus] .= v_a
+        t1sx .= arrays.x
+        adj_t1sx .= 0.0
+        t1sF .= 0.0
+        adj_t1sF .= 0.0
+    elseif isa(type, ControlJacobian)
+        x[1:nvbus] .= v_m
+        x[nvbus+1:nvbus+ninj] .= pinj
+        t1sx .= arrays.x
+        adj_t1sx .= 0.0
+        t1sF .= 0.0
+        adj_t1sF .= 0.0
+    else
+        error("Unsupported Jacobian structure")
+    end
+    t1sF .= 0.0
+    adj_t1sF .= lambda
+    adj_t1sx .= 0.0
+    seed_kernel!(arrays.t1sseeds, arrays.varx, arrays.t1svarx, nbus)
+    if isa(type, StateJacobian)
+        residual_adj_polar!(
+            t1sF, adj_t1sF,
+            view(t1sx, 1:nvbus),
+            view(adj_t1sx, 1:nvbus),
+            view(t1sx, nvbus+1:2*nvbus),
+            view(adj_t1sx, nvbus+1:2*nvbus),
+            ybus_re, ybus_im,
+            pinj, zeros(eltype(pinj), length(pinj)),
```

Comment:
Why instantiating a new array there?

Reply:
Those are the adjoints. They have to be zero.